staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go

```diff
@@ -175,26 +178,49 @@ type batchBackend struct {
 	// Receiving maxBatchSize events will always trigger a send, regardless of
 	// if this amount of time has been reached.
 	maxBatchWait time.Duration
+
+	// Channel to signal that the main routine stopped and no new requests will
+	// be sent to the webhook.
+	shutdownCh chan struct{}
+
+	// Mutex to lock on until all requests to the webhook were successfully sent.
+	reqMutex sync.RWMutex
 }
 
 func (b *batchBackend) Run(stopCh <-chan struct{}) error {
-	f := func() {
+	collect := func() []auditinternal.Event {
+		t := time.NewTimer(b.maxBatchWait)
+		defer t.Stop() // Release ticker resources
+
+		return b.collectEvents(stopCh, t.C)
+	}
+	send := func(events []auditinternal.Event) {
 		// Recover from any panics caused by this method so a panic in the
 		// goroutine can't bring down the main routine.
 		defer runtime.HandleCrash()
 
-		t := time.NewTimer(b.maxBatchWait)
-		defer t.Stop() // Release ticker resources
-
-		b.sendBatchEvents(stopCh, t.C)
+		b.sendBatchEvents(events)
 	}
 
 	go func() {
 		for {
-			f()
+			send(collect())
```

Comment:
can this loop panic? If it does we will lock up on shutdown. We should have a panic handler.